### ISSUE TITLE
Disabling file concatenation

### ISSUE THREAD

**Issue by dmail** [rollup/rollup#3882]
### Context

For some projects I would like to use rollup to benefit from tree-shaking and url hashing without generating chunks (disable concatenation of files). From my understanding `preserveModules` is the option to use for that situation. According to my tests `preserveModules` behaviour is not what I expected.

The following code illustrates what I have in mind.

```js
import { rollup } from "rollup"
import { fileURLToPath } from "url"

const start = async () => {
  const projectDirectoryUrl = new URL("./", import.meta.url)

  const fileContents = {
    "./main.js": `import { value } from "./dep.js"
console.log(value)`,
    "./dep.js": `export const value = 42`,
  }

  const rollupBundle = await rollup({
    input: ["./main.js"],
    plugins: [
      {
        resolveId: (id) => {
          return id
        },
        load: (id) => {
          return fileContents[id]
        },
      },
    ],
  })

  await rollupBundle.write({
    dir: fileURLToPath(`${projectDirectoryUrl}dist`),
    preserveModules: true,
    chunkFileNames: `[name]-[hash].js`,
  })
}

start()
```

### Expected Behavior / Situation

Rollup write files with hash in their names such as 

```
dist/main-25e95a00.js
dist/dep-5dd85f23.js
```

### Actual Behavior / Situation

Rollup write files without hash under `_virtual` directory.

```
dist/_virtual/main.js
dist/_virtual/dep.js
```

### Modification Proposal

Allow to disable chunks using `preserveModules` options or something else and update import paths to get hash in urls.

**Comment by lukastaegert** [comment]
This has been discussed in the past in different contexts, see e.g. #3684. The way forward will be to fully respect the `entryFileNames` and `chunkFileNames` options when preserving modules. Unfortunately due to some choices in the current implementation, this will be a breaking change and is therefore marked for an upcoming 3.0 release. No due date for that, unfortunately.

**Comment by dmail** [comment]
Thank you for the feedback 👍. I have no special urgency around this need, it's more for experimentation.
I saw the issue about node_modules but could not follow what was the root issue in that thread, sorry if that's a duplicate.

**Comment by lukastaegert** [comment]
It is now kind of tracked here: https://github.com/rollup/rollup/projects/3

**Comment by dmail** [comment]
By the way I found a way to disable concatenation by accident. 

In short, inside the "load" hook, I emit a chunk for every js file.

```js
async load(id) {
  if (id.endsWith('.js')) {
    this.emit({type: "chunk", id })
  }
}
```



.

**Comment by tirrth** [comment]
@dmail Your fix doesn't seem to be working for me. Does anyone's got some sorta workaround for this?

**Comment by lukastaegert** [comment]
Resolved in #4565

**Comment by lukastaegert** [comment]
This issue has been resolved via #4565 as part of rollup@3.0.0-3. Note that this is a pre-release, so to test it, you need to install Rollup via `npm install rollup@3.0.0-3` or `npm install rollup@beta`. It will likely become part of a regular release later.

**Comment by lukastaegert** [comment]
This issue has been resolved via #4565 as part of rollup@3.0.0-3. Note that this is a pre-release, so to test it, you need to install Rollup via `npm install rollup@3.0.0-3` or `npm install rollup@beta`. It will likely become part of a regular release later.

**Comment by lukastaegert** [comment]
This issue has been resolved via #4565 as part of rollup@3.0.0-4. Note that this is a pre-release, so to test it, you need to install Rollup via `npm install rollup@3.0.0-4` or `npm install rollup@beta`. It will likely become part of a regular release later.
